SpatialGDK/Documentation/content/non-unreal-layers.md

```diff
@@ -47,20 +47,17 @@ There is a basic example in the _Examples_ section below. For more examples of h
 
 >**Note:** Your external SpatialOS components must have an ID between 1000 and 2000 to be registered by the pipeline.
 
-You set up your game to receive component updates through implementing callbacks for specified SpatialOS component IDs and operation types. The GDK detects your callbacks when it’s initialized and uses them for all received data relating to components with IDs in the range 1000 to 2000.
-To create callbacks, you need to create a class derived from the GDK `UOpCallbackTemplate` base class and implement `GetComponentId` and one or more of the callback methods. These callback methods are parameterized by the following network operation types:
+You set up your game to receive network operations using the `SpatialNetDriver::AddOpCallback` function. This function is paramaterised with a SpatialOS component ID and a callback function reference with the following type:
 
-* `virtual void OnAddComponent(const Worker_AddComponentOp& Op) {}`
-* `virtual void OnRemoveComponent(const Worker_RemoveComponentOp& Op) {}`
-* `virtual void OnComponentUpdate(const Worker_ComponentUpdateOp& Op) {}`
-* `virtual void OnAuthorityChange(const Worker_AuthorityChangeOp& Op) {}`
-* `virtual void OnCommandRequest(const Worker_CommandRequestOp& Op) {}`
-* `virtual void OnCommandResponse(const Worker_CommandResponseOp& Op) {}`
+`const TFunction<void(Worker_ComponentId, const Worker_Op*)>`
 
+where `Worker_ComponentId` and `Worker_Op` are types defined in the [Worker SDK in C’s API](https://docs.improbable.io/reference/latest/capi/reference).
 
-The `UOpCallbackTemplate` base class contains protected references to the `UWorld` and GDK `StaticComponentView` enabling you to write more meaningful callbacks.
+You'll need to register your callbacks before the Unreal worker connects to the SpatialOS runtime to avoid missing any operations.
```

Comment:
I know this has been signed off for mvp, but this sort of thing worries me :(

Reply:
Potential noob question :) Why does this worry you? because you expect users to want to add callbacks during the runtime of a game?

Reply:
It's introducing quite a narrow contract that is easy to use incorrectly. It also implies that users will have to care about exact op semantics and interpret them correctly.

Reply:
It's not wrong, it's just narrowing a contract because it was easy to make rather than because it makes something faster or helps a user in some broader context.